openedx_tagging/core/tagging/models.py

```diff
@@ -1,39 +1,428 @@
 """ Tagging app data models """
+from typing import List, Type
+
 from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+from openedx_learning.lib.fields import MultiCollationTextField, case_insensitive_char_field
+
+# Maximum depth allowed for a hierarchical taxonomy's tree of tags.
+TAXONOMY_MAX_DEPTH = 3
+
+# Ancestry of a given tag; the Tag.value fields of a given tag and its parents, starting from the root.
+# Will contain 0...TAXONOMY_MAX_DEPTH elements.
+Lineage = List[str]
+
+
+class Tag(models.Model):
+    """
+    Represents a single value in a list or tree of values which can be applied to a particular Open edX object.
+
+    Open edX tags are "name:value" pairs which can be applied to objects like content libraries, units, or people.
+    Tag.taxonomy.name provides the "name" and the Tag.value provides the "value".
+    (And an ObjectTag links a Tag with an object.)
+    """
+
+    id = models.BigAutoField(primary_key=True)
+    taxonomy = models.ForeignKey(
+        "Taxonomy",
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        help_text=_("Namespace and rules for using a given set of tags."),
+    )
+    parent = models.ForeignKey(
+        "self",
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        related_name="children",
+        help_text=_(
+            "Tag that lives one level up from the current tag, forming a hierarchy."
+        ),
+    )
+    value = case_insensitive_char_field(
+        max_length=500,
+        help_text=_(
+            "Content of a given tag, occupying the 'value' part of the key:value pair."
+        ),
+    )
+    external_id = case_insensitive_char_field(
+        max_length=255,
+        null=True,
+        blank=True,
+        help_text=_(
+            "Used to link an Open edX Tag with a tag in an externally-defined taxonomy."
+        ),
+    )
+
+    class Meta:
+        index_together = ["taxonomy", "value"]
```

Comment:
This style is deprecated in 4.2. Please use [indexes](https://docs.djangoproject.com/en/4.2/ref/models/options/#django.db.models.Options.indexes) instead.